tests/api/test_dnf_logging.py

```diff
@@ -0,0 +1,64 @@
+# -*- coding: utf-8 -*-
+
+
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+import dnf
+
+import logging
+from .common import TestCase
+
+
+class DnfLoggingApiTest(TestCase):
+    def test_levels(self):
+        self.assertHasAttr(dnf.logging, "SUPERCRITICAL")
+        self.assertHasType(dnf.logging.SUPERCRITICAL, int)
+
+        self.assertHasAttr(dnf.logging, "CRITICAL")
+        self.assertHasType(dnf.logging.CRITICAL, int)
+
+        self.assertHasAttr(dnf.logging, "ERROR")
+        self.assertHasType(dnf.logging.ERROR, int)
+
+        self.assertHasAttr(dnf.logging, "WARNING")
+        self.assertHasType(dnf.logging.WARNING, int)
+
+        self.assertHasAttr(dnf.logging, "INFO")
+        self.assertHasType(dnf.logging.INFO, int)
+
+        self.assertHasAttr(dnf.logging, "DEBUG")
+        self.assertHasType(dnf.logging.DEBUG, int)
+
+        self.assertHasAttr(dnf.logging, "DDEBUG")
+        self.assertHasType(dnf.logging.DDEBUG, int)
+
+        self.assertHasAttr(dnf.logging, "SUBDEBUG")
+        self.assertHasType(dnf.logging.SUBDEBUG, int)
+
+        self.assertHasAttr(dnf.logging, "TRACE")
+        self.assertHasType(dnf.logging.TRACE, int)
+
+    def test_logging_level_names(self):
+        # Level names added in dnf logging initialization
+        self.assertTrue(logging.getLevelName(dnf.logging.DDEBUG) == "DDEBUG")
+        self.assertTrue(logging.getLevelName(dnf.logging.SUBDEBUG) == "SUBDEBUG")
+        self.assertTrue(logging.getLevelName(dnf.logging.TRACE) == "TRACE")
+
+    def test_dnf_logger(self):
+        # This doesn't really test much since python allows getting any logger,
+        # at least check if it has handlers (setup in dnf)
+        logger = logging.getLogger('dnf')
+        self.assertTrue(len(logger.handlers) > 0)
+
+    def test_dnf_rpm_logger(self):
+        # This doesn't really test much since python allows getting any logger,
+        # at least check if it has handlers (setup in dnf)
+        logger = logging.getLogger('dnf.rpm')
+        self.assertTrue(len(logger.handlers) > 0)
+
+    def test_dnf_plugin_logger(self):
+        # This doesn't really test much since python allows getting any logger,
+        # dnf.plugin logger also doens't have any setup in dnf, so it cannot be
+        # differentiated from any other random logger.
+        logger = logging.getLogger('dnf.plugin')
```

Comment:
I think the test for "dnf" logger can stay as it is. I haven't found any better way to test if logger exists and testing it has some handlers seems good enough.

I was wondering how the "dnf.plugin" logger even works when it has no handlers and it turns out there is a hierarchy in python loggers given by the "." separator. So, it leads me to an idea that for the "dnf.rpm" and "dnf.plugin" loggers, it would make sense to test that either they have some handlers or the messages are propagated to the parent (attribute ``propagate`` is True). I don't know if these are the only ways to handle the log messages, but we can change the test later if we find the need for it.

Reply:
I have added the tests for `propagate`, "dnf.rpm" has it false:
https://github.com/rpm-software-management/dnf/blob/639c10e59c8b5afec62dab25fcb3d08f6192b585/dnf/logging.py#L189
and for "dnf.plugin" it is true by default.

I wanted to add the test for handlers to "dnf.plugin" as well but it doesn't work, I think that it doesn't show all parents handlers only if it would have some of its own.

Reply:
I meant to test together that either it has handlers or ``propagate=True``, i.e. add both into one assert. Because I think that the compound condition should always be true for both of them, as it means that the logger is somehow handled.

Reply:
I see now, so while the condition will be true by default for any logger it makes sure that it is at least somehow handled like you say. That makes sense to me, updated.